graspologic/embed/n2v.py

```diff
@@ -0,0 +1,402 @@
+        Dimensionality of the word vectors. Default is 128.
+    window_size : int
+        Maximum distance between the current and predicted word within a sentence. Default is 10.
+    workers : int
+        Use these many worker threads to train the model. Default is 8.
+    iterations : int
+        Number of epochs in stochastic gradient descent (SGD)
+    interpolate_walk_lengths_by_node_degree : bool
+        Use a dynamic walk length that corresponds to each nodes
+        degree. If the node is in the bottom 20 percentile, default to a walk length of 1. If it is in the top 10
+        percentile, use walk_length. If it is in the 20-80 percentiles, linearly interpolate between 1 and walk_length.
+        This will reduce lower degree nodes from biasing your resulting embedding. If a low degree node has the same
+        number of walks as a high degree node (which it will if this setting is not on), then the lower degree nodes
+        will take a smaller breadth of random walks when compared to the high degree nodes. This will result in your
+        lower degree walks dominating your higher degree nodes.
+
+    Returns
+    Tuple[np.ndarray, np.ndarray]
+        A tuple containing a matrix, with each row index corresponding to the embedding for each node. The tuple
+        also contains a vector containing the corresponding vertex labels for each row in the matrix.
+        The matrix and vector are positionally correlated.
+    """
+    node2vec_graph = _Node2VecGraph(graph, return_hyperparameter, inout_hyperparameter)
+
+    logging.info(
+        f"Starting preprocessing of transition probabilities on graph with {str(len(graph.nodes()))} nodes and "
+        f"{str(len(graph.edges()))} edges"
+    )
+
+    start = time.time()
+    logging.info(f"Starting at time {str(start)}")
+
+    node2vec_graph._preprocess_transition_probabilities()
+
+    logging.info(f"Simulating walks on graph at time {str(time.time())}")
+    walks = node2vec_graph._simulate_walks(
+        num_walks, walk_length, interpolate_walk_lengths_by_node_degree
+    )
+
+    logging.info(f"Learning embeddings at time {str(time.time())}")
+    model = _learn_embeddings(walks, dimensions, window_size, workers, iterations)
+
+    end = time.time()
+    logging.info(
+        f"Completed. Ending time is {str(end)} Elapsed time is {str(start - end)}"
+    )
+
+    return model.wv.vectors, model.wv.index2word
+
+
+def _learn_embeddings(
+    walks: List[Any], dimensions: int, window_size: int, workers: int, iterations: int
+):
+    """
+    Learn embeddings by optimizing the skip-gram objective using SGD.
+    """
+    from gensim.models import Word2Vec
+
+    walks = [list(map(str, walk)) for walk in walks]
+
+    # Documentation - https://radimrehurek.com/gensim/models/word2vec.html
+    model = Word2Vec(
+        walks,
+        size=dimensions,
+        window=window_size,
+        min_count=0,
+        sg=1,  # Training algorithm: 1 for skip-gram; otherwise CBOW
```

Comment:
where'd this come from? (the comment here)

Reply:
like 99% sure from Zeinab - she wrote all of this originally, with @Nyecarr modifying it after the fact